hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h

```diff
@@ -226,10 +217,36 @@ namespace hazelcast {
                                                 }));
                     }
 
+                    void do_write(const std::shared_ptr<connection::Connection> connection,
+                               const std::shared_ptr<spi::impl::ClientInvocation> invocation) {
+                        auto handler = [connection, invocation, this](const boost::system::error_code &ec,
+                                                                      std::size_t bytes_written) {
+                            this->outbox_.pop_front();
+
+                            if (ec) {
+                                auto message = (boost::format{
+                                        "Error %1% during invocation write for %2% on connection %3%"} %
+                                                ec % *invocation % *connection).str();
+                                connection->close(message);
+                            } else {
+                                // update the connection write time
+                                connection->last_write_time(std::chrono::steady_clock::now());
+
+                                if (!this->outbox_.empty()) {
+                                    do_write(connection, invocation);
```

Comment:
This handler can't run in parallel with `socket.strand.post()` handler above, right? Are they executed in the same thread?

Reply:
We execute handler using this wrapper `socket_strand_.wrap(handler)` which guarantees that the handler is executed using the same strand. Strand guarantees that the codes using the same strands can not run in parallel, they will be sequential through the strand. see asio strand concept. 

For our implementation, yes they are at the same io thread. As a matter of fact we only have a single io thread.